addOns/authhelper/src/main/java/org/zaproxy/addon/authhelper/HeaderBasedSessionManagementPanel.java

```diff
@@ -0,0 +1,174 @@
+        return LayoutHelper.getGBC(
+                x,
+                y,
+                1,
+                1.0,
+                1.0,
+                GridBagConstraints.BOTH,
+                GridBagConstraints.NORTHWEST,
+                STD_INSETS);
+    }
+
+    private Pair<ZapTextField, ZapTextField> addRow() {
+        Pair<ZapTextField, ZapTextField> pair = new Pair<>(new ZapTextField(), new ZapTextField());
+        fields.add(pair);
+        this.add(pair.first, getGBC(0, fields.size()));
+        this.add(pair.second, getGBC(1, fields.size()));
+        pair.second.addKeyListener(
+                new KeyAdapter() {
+                    @Override
+                    public void keyReleased(KeyEvent e) {
+                        super.keyReleased(e);
+                        manageRows();
+                    }
+                });
+        return pair;
+    }
+
+    private void manageRows() {
+        Pair<ZapTextField, ZapTextField> last = fields.get(fields.size() - 1);
+        if (!last.second.getText().isBlank()) {
+            // Last field no longer blank, add another one
+            EventQueue.invokeLater(
+                    () -> {
+                        addRow();
+                        this.revalidate();
+                        this.repaint();
+                    });
+            return;
+        }
+        if (fields.size() > 2 && last.first.getText().isBlank()) {
+            Pair<ZapTextField, ZapTextField> prev = fields.get(fields.size() - 2);
+            if (prev.first.getText().isBlank() && prev.second.getText().isBlank()) {
+                // Last 2 rows now blank - remove the last one
+                EventQueue.invokeLater(
+                        () -> {
+                            this.remove(last.first);
+                            this.remove(last.second);
+                            fields.remove(last);
+                            this.revalidate();
+                            this.repaint();
+                        });
+            }
+        }
+    }
+
+    private Pair<ZapTextField, ZapTextField> getNextFields() {
+        for (Pair<ZapTextField, ZapTextField> pair : fields) {
+            if (pair.first.getText().isBlank() && pair.second.getText().isBlank()) {
+                return pair;
+            }
+        }
+        return addRow();
+    }
+
+    private void clearFields() {
+        for (Pair<ZapTextField, ZapTextField> pair : fields) {
+            this.remove(pair.first);
+            this.remove(pair.second);
+        }
+        fields.clear();
+        addRow();
+        addRow();
+        this.revalidate();
+        this.repaint();
+    }
+
+    public void setHeaders(List<Pair<String, String>> headers) {
+        clearFields();
+
+        for (Pair<String, String> header : headers) {
+            Pair<ZapTextField, ZapTextField> fields = getNextFields();
+            fields.first.setText(header.first);
+            fields.second.setText(header.second);
```

Comment:
Ideally we should discard the edits after setting the text to not undo to empty text field.